brickstrap-image-drivers.sh

```diff
@@ -0,0 +1,151 @@
+        "$1"=bootroot:vfat:ext4:${IMAGE_FILE_SIZE}:48M:mbr \
+        part-set-mbr-id /dev/sda 1 0x0b : \
+        set-label /dev/sda2 EV3_FILESYS : \
+        mount /dev/sda2 / : \
+        tar-in "$(br_tarball_path)" / : \
+        mkdir-p /media/mmc_p1 : \
+        mount /dev/sda1 /media/mmc_p1 : \
+        glob mv /boot/flash/* /media/mmc_p1/ : \
+
+    # Hack to set the volume label on the vfat partition since guestfish does
+    # not know how to do that. Must be null padded to exactly 11 bytes.
+    echo -e -n "EV3_BOOT\0\0\0" | \
+        dd of="$1" bs=1 seek=32811 count=11 conv=notrunc >/dev/null 2>&1
+}
+
+#
+# Convert megabytes to sectors. Sectors are assumed to be 512 bytes big.
+# $1 is size in megabytes.
+function brp_to_sector()
+{
+    #echo $(( $1 * 1024 * 1024 / 512 ))
+    echo $(( $1 * 2048 ))
+}
+
+#
+# Create a disk image with MBR partition table and 4 partitions. There are two
+# identical rootfs partitions to allow for failover and live upgrades.
+# ------------------------------------------------------------------------------
+# part | label              | mount point | fs   | size
+# ------------------------------------------------------------------------------
+#    1 | ${BOOT_PART_NAME}  | /boot/flash | VFAT | 48MB
+#    2 | ${ROOT_PART_NAME}1 | /           | ext4 | ${ROOT_PART_SIZE}
+#    3 | ${ROOT_PART_NAME}2 | /mnt/root2  | ext4 | ${ROOT_PART_SIZE}
+#    4 | ${DATA_PART_NAME}  | /var        | ext4 | ${IMAGE_FILE_SIZE} -
+#      |                    |             |      | 2 * ${ROOT_PART_SIZE} - 48MB
+# ------------------------------------------------------------------------------
+#
+# $1: path to the image file to generate.
+#
+function brp_image_drv_redundant_rootfs_w_data()
+{
+    [ $# -eq 1 -a -n "$1" ] && guestfish -N "$1"=disk:${IMAGE_FILE_SIZE} -- \
+        part-init /dev/sda mbr : \
+        part-add /dev/sda primary 0 $(brp_to_sector 48) : \
+        part-add /dev/sda primary $(brp_to_sector 48) $(brp_to_sector ${ROOT_PART_SIZE}) : \
+        part-add /dev/sda primary $(brp_to_sector ${ROOT_PART_SIZE}) $(brp_to_sector $(( 2 * ${ROOT_PART_SIZE} ))) : \
+        part-add /dev/sda primary $(brp_to_sector $(( 2 * ${ROOT_PART_SIZE} ))) -1 : \
+        part-set-mbr-id /dev/sda 1 0x0b : \
+        mkfs fat /dev/sda1 : \
+        set-label /dev/sda1 ${BOOT_PART_NAME} : \
+        mkfs ext4 /dev/sda2 : \
+        set-label /dev/sda2 ${ROOT_PART_NAME}1 : \
+        mkfs ext4 /dev/sda3 : \
+        set-label /dev/sda3 ${ROOT_PART_NAME}2 : \
+        mkfs ext4 /dev/sda4 : \
+        set-label /dev/sda4 ${DATA_PART_NAME} : \
+        mkdir-p /boot/flash : \
+        mount /dev/sda1 /boot/flash : \
+        mount /dev/sda2 / : \
+        mkdir-p /mnt/root2
+        mount /dev/sda3 /var : \
+        mkdir-p /var : \
+        mount /dev/sda4 /var : \
+        tar-in "$(br_tarball_path)" / : \
+        umount /boot/flash : \
+        umount /var : \
+        glob cp-a /* /mnt/root2/ : \
```

Comment:
Since there is a `\` at the end of the line here, it is important that the next line is blank. It's been a while since I messed with this, but I remember guestfish being weird, so I don't think removing the trailing  `: \` is an option.